internal/metrics/prometheus.go

```diff
@@ -0,0 +1,159 @@
+// ones. Because of that, we have to use its global registry as CR doesn't allow injecting a custom one.
+// Upstream issue regarding this: https://github.com/kubernetes-sigs/controller-runtime/issues/210.
+type GlobalCtrlRuntimeMetricsRecorder struct{}
+
+var _ Recorder = &GlobalCtrlRuntimeMetricsRecorder{}
+
+func NewGlobalCtrlRuntimeMetricsRecorder() *GlobalCtrlRuntimeMetricsRecorder {
+	return &GlobalCtrlRuntimeMetricsRecorder{}
+}
+
+func (r *GlobalCtrlRuntimeMetricsRecorder) RecordKonnectEntityOperationSuccess(
+	operationType KonnectEntityOperation, entityType string, duration time.Duration) {
+	r.recordKonnectEntityOperationCount(operationType, entityType, true, 0)
+	r.recordKonnectEntityOperationDuration(operationType, entityType, true, 0, duration)
+}
+
+func (r *GlobalCtrlRuntimeMetricsRecorder) RecordKonnectEntityOperationFailure(
+	operationType KonnectEntityOperation, entityType string, duration time.Duration, statusCode int) {
+	r.recordKonnectEntityOperationCount(operationType, entityType, false, statusCode)
+	r.recordKonnectEntityOperationDuration(operationType, entityType, false, statusCode, duration)
+}
+
+func (r *GlobalCtrlRuntimeMetricsRecorder) recordKonnectEntityOperationCount(
+	operationType KonnectEntityOperation, entityType string, success bool, statusCode int,
+) {
+	labels := konnectEntityOperationLabels(operationType, entityType, success, statusCode)
+	konnectEntityOperationCount.With(labels).Inc()
+}
+
+func (r *GlobalCtrlRuntimeMetricsRecorder) recordKonnectEntityOperationDuration(
+	operationType KonnectEntityOperation, entityType string, success bool, statusCode int, duration time.Duration,
+) {
+	labels := konnectEntityOperationLabels(operationType, entityType, success, statusCode)
+	konnectEntityOperationDuration.With(labels).Observe(duration.Seconds())
+}
+
+func konnectEntityOperationLabels(operationType KonnectEntityOperation, entityType string, success bool, statusCode int,
+) prometheus.Labels {
+	labels := prometheus.Labels{
+		KonnectEntityOperationTypeKey: string(operationType),
+		KonnectEntityTypeKey:          entityType,
+	}
+	if success {
+		labels[SuccessKey] = SuccessTrue
+		labels[StatusCodeKey] = ""
+	} else {
+		labels[SuccessKey] = SuccessFalse
+		labels[StatusCodeKey] = strconv.Itoa(statusCode)
+	}
+	return labels
+}
+
+func init() {
+	allMetrics := []prometheus.Collector{
+		konnectEntityOperationCount,
+		konnectEntityOperationDuration,
+	}
+	for _, m := range allMetrics {
+		ctrlmetrics.Registry.MustRegister(m)
+	}
+}
+
+type MockRecorder struct{}
+
+var _ Recorder = &MockRecorder{}
+
+func (m *MockRecorder) RecordKonnectEntityOperationSuccess(
+	operationType KonnectEntityOperation, entityType string, duration time.Duration) {
+}
+
+func (m *MockRecorder) RecordKonnectEntityOperationFailure(
+	operationType KonnectEntityOperation, entityType string, duration time.Duration, statusCode int) {
+}
```

Comment:
I believe mocks should live in a separate directory, nested under `test/` top level dir just like we did recently in KIC: https://github.com/Kong/kubernetes-ingress-controller/pull/6868

Reply:
Moving all the mocks has a much larger scope than this PR. I've created a new issue for moving mocks: https://github.com/Kong/gateway-operator/issues/955.